File: solutions/03-Rucksack_Reorganization/part-1.py
# --- Day 3: Rucksack Reorganization [Part One] ---


import string


def halves(item: str) -> tuple[str, str]:
    halfway = int(len(item) / 2)
    return item[:halfway], item[halfway:]


def find_duplicate(halves: tuple[str, str]) -> str:
    for char in halves[0]:
        if char in halves[1]:
            return char


SCORE = {char: i + 1 for i, char in enumerate(string.ascii_letters)}

# Load Data
raw = open("contents.txt").read()
rucksacks = [halves(item) for item in raw.split("\n")]

# Find Duplicates
duplicates = [find_duplicate(rucksack) for rucksack in rucksacks]

# Add Scores
total_score = sum([SCORE[char] for char in duplicates])
print(f"Total Score: {total_score}")

File: solutions/04-Camp_Cleaning/part-1.js
// --- Day 4: Camp Cleaning [Part One] ---

const fs = require("fs");

function extractRange(assignment) {
  return assignment
    .split(",")
    .map((section) => section.split("-").map((value) => parseInt(value)));
}

function isWithin(range1, range2) {
  return range1[0] >= range2[0] && range1[1] <= range2[1];
}

function checkContains(ranges) {
  const [range1, range2] = ranges;
  return isWithin(range1, range2) || isWithin(range2, range1);
}

// Load Data
const raw = fs.readFileSync("sections.txt", "utf8");
const assignments = raw.split("\n");

// Extract Ranges
const sectionRanges = assignments.map((assignment) => extractRange(assignment));

// Find Totally Overlapped Ranges
const contained = sectionRanges.map((ranges) => checkContains(ranges));

// Count Score
totalScore = contained.filter(Boolean).length;
console.log(`Total Score: ${totalScore}`);
